src/FileFormats/format_flatbuffer_opl3.cpp

```diff
@@ -0,0 +1,206 @@
+    if(!file.open(QIODevice::ReadOnly))
+        return FfmtErrCode::ERR_NOFILE;
+    
+    QByteArray blob = file.readAll();
+    file.close();
+
+    // Get a pointer to the root object inside the buffer.
+    auto opl3Bank = GetOpl3Bank(blob.constData());
+    auto banks = opl3Bank->banks();
+
+    // count banks types
+    for (uint8_t i = 0; i < banks->Length(); i++) {
+        auto bnk = banks->Get(i);
+        switch (bnk->type()) {
+            case BankType_Melodic:
+                count_melodic_banks++;
+                break;
+            case BankType_Percussion:
+                count_percusive_banks++;
+                break;
+            default:
+                return FfmtErrCode::ERR_BADFORMAT;
+        }
+    }
+
+    bank.reset(count_melodic_banks, count_percusive_banks);
+
+    bank.deep_vibrato = ((opl3Bank->oplTV() >> 0) & 0x01);
+    bank.deep_tremolo = ((opl3Bank->oplTV() >> 1) & 0x01);
+    switch (opl3Bank->volumeModel()) {
+        case VolumeModel_Auto:
+            bank.volume_model = 0;
+            break;
+        case VolumeModel_Generic:
+            bank.volume_model = 1;
+            break;
+        case VolumeModel_Native:
+            bank.volume_model = 2;
+            break;
+        case VolumeModel_DMX:
+            bank.volume_model = 3;
+            break;
+        case VolumeModel_Apogee:
+            bank.volume_model = 4;
+            break;
+        case VolumeModel_Win9x:
+            bank.volume_model = 5;
+            break;
+        default:
+            bank.volume_model = 0;
+            break;        
+    }
+    bank.volume_model = opl3Bank->volumeModel();
+
+    uint16_t current_melodic_bank     = 0;
+    uint16_t current_percusive_bank   = 0;
+    FmBank::MidiBank* bankMeta = NULL;
+
+    for(uint8_t i = 0; i < banks->Length(); i++) {
+        auto bnk = banks->Get(i);
+        auto bankType = bnk->type();
+
+        switch (bankType) {
+            case BankType_Melodic:
+                bankMeta = &bank.Banks_Melodic[current_melodic_bank];
+                break;
+            case BankType_Percussion:
+                bankMeta = &bank.Banks_Percussion[current_percusive_bank];
+                break;
+        }
+
+        bankMeta->lsb = bnk->bankLSB();
+        bankMeta->msb = bnk->bankMSB();
+        strncpy(bankMeta->name, bnk->name()->c_str(), 32);
+
+        auto instruments = bnk->instruments();
+        uint8_t instrumentCount = instruments->Length();
+
+        for (uint8_t j = 0; j < instrumentCount; j++) {
+            auto instrument = instruments->Get(j);
+
+            FmBank::Instrument* ins = NULL;
+            switch (bankType) {
+                case BankType_Melodic:
+                    ins = &bank.Ins_Melodic[current_melodic_bank * 128 + instrument->program()];
+                    break;
+                case BankType_Percussion:
+                    ins = &bank.Ins_Percussion[current_percusive_bank * 128 + instrument->percussionKey()];
+                    break;
```

Comment:
Percussion key is NOT the slot where to store instrument. Use program() too!
Percussion key means actual MIDI note will be played when playing this drum. No matter from the slot where it is stored.